src/main/java/com/google/devtools/build/lib/remote/RemoteOutputService.java

```diff
@@ -231,4 +246,293 @@ public void checkActionFileSystemForLostInputs(FileSystem actionFileSystem, Acti
       remoteFileSystem.checkForLostInputs(action);
     }
   }
+
+  @Override
+  public void registerPostExecutionTask(ActionExecutionMetadata action, Cancellable task) {
+    // We don't expect to have multiple post-execution tasks for the same action registered at the
+    // same time.
+    postExecutionTasks.merge(
+        action,
+        task,
+        (oldTask, newTask) -> {
+          throw new IllegalStateException(
+              "Attempted to register multiple post-execution tasks for %s: %s and %s"
+                  .formatted(action, oldTask, newTask));
+        });
+  }
+
+  @Override
+  public void cancelPostExecutionTasks(ActionExecutionMetadata action) throws InterruptedException {
+    Cancellable task = postExecutionTasks.remove(action);
+    if (task != null) {
+      task.cancel();
+    }
+  }
+
+  @Override
+  public RewoundActionSynchronizer createRewoundActionSynchronizer(boolean rewindingEnabled) {
+    if (rewindingEnabled && actionInputFetcher != null) {
+      return new RemoteRewoundActionSynchronizer();
+    }
+    return RewoundActionSynchronizer.NOOP;
+  }
+
+  /**
+   * A {@link RewoundActionSynchronizer} implementation for Bazel's remote filesystem, which is
+   * backed by actual files on disk and requires synchronization to ensure that action outputs
+   * aren't deleted while they are being read.
+   */
+  final class RemoteRewoundActionSynchronizer implements RewoundActionSynchronizer {
+    // A single coarse lock is used to synchronize rewound actions (writers) and both rewound and
+    // non-rewound actions (readers) as long as no rewound action has attempted to prepare for its
```

Comment:
@coeuvre Regarding the optimization potential we discussed privately: If we can be sure that a certain action writes its outputs atomically and doesn't need them to be deleted (say, if we now the action runs with remote execution, where we can arrange for this), it would not need to acquire the write lock. If all actions are of this type (as they would be at Google), the fine-grained lock would never be inflated.